### ISSUE TITLE
Flipping navigator causes incorrect navigator display region rotation

### ISSUE THREAD

**Issue by jbakarich** [openseadragon/openseadragon#2612]
### Description
- It seems that the `$.Navigator.setFlip()` function overrides or incorrectly sets the `rotate()` value of the displayRegion element
- Setting flip, then rotating rotates the display region of the navigator correctly
- Setting rotation, then flipping causes the display region to be flipped correctly but have an incorrect rotation

**Example:**

![image](https://github.com/user-attachments/assets/fc779524-c94c-4010-9faf-f2218de55730)
- Note the rotation of the display region in the navigator vs the FOV shown
- This was achieved by rotating, then flipping the TiledImage ([see linked sandbox](https://codesandbox.io/p/sandbox/openseadragon-sandbox-forked-3yfr86?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsw0ywvg00072v6hngtlh49h%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsw0ywvg00032v6hd9h3xqfw%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsw0ywvg00042v6h99qbxple%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsw0ywvg00062v6hbwdu5ral%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsw0ywvg00032v6hd9h3xqfw%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsw0ywvg00022v6hauu15b2t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A37%252C%2522startColumn%2522%253A15%252C%2522endLineNumber%2522%253A37%252C%2522endColumn%2522%253A15%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%257D%255D%252C%2522id%2522%253A%2522clsw0ywvg00032v6hd9h3xqfw%2522%252C%2522activeTabId%2522%253A%2522clsw0ywvg00022v6hauu15b2t%2522%257D%252C%2522clsw0ywvg00062v6hbwdu5ral%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsw0ywvg00052v6h080o8ppl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsw0ywvg00062v6hbwdu5ral%2522%252C%2522activeTabId%2522%253A%2522clsw0ywvg00052v6h080o8ppl%2522%257D%252C%2522clsw0ywvg00042v6h99qbxple%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsw0ywvg00042v6h99qbxple%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D&workspaceId=58fd793c-a178-4c0e-8ac3-54df60751a6e)) 

**Correct Example:**

![image](https://github.com/user-attachments/assets/cab3039e-89f3-4318-9966-9897fcedc376)
- Note that the display region accurately reflects the FOV

### Version
- 5.0.0 (but present in 4.x) 

### Steps to repro

- Add a `TiledImage`
- Rotate then flip via the `viewport.setFlip` and `viewport.setRotation` methods
- Note the incorrect position of the display region

### Sandbox
[Link to sandbox with example](https://codesandbox.io/p/sandbox/openseadragon-sandbox-forked-3yfr86?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsw0ywvg00072v6hngtlh49h%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsw0ywvg00032v6hd9h3xqfw%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsw0ywvg00042v6h99qbxple%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsw0ywvg00062v6hbwdu5ral%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B40%252C60%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsw0ywvg00032v6hd9h3xqfw%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsw0ywvg00022v6hauu15b2t%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A37%252C%2522startColumn%2522%253A15%252C%2522endLineNumber%2522%253A37%252C%2522endColumn%2522%253A15%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%257D%255D%252C%2522id%2522%253A%2522clsw0ywvg00032v6hd9h3xqfw%2522%252C%2522activeTabId%2522%253A%2522clsw0ywvg00022v6hauu15b2t%2522%257D%252C%2522clsw0ywvg00062v6hbwdu5ral%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsw0ywvg00052v6h080o8ppl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsw0ywvg00062v6hbwdu5ral%2522%252C%2522activeTabId%2522%253A%2522clsw0ywvg00052v6h080o8ppl%2522%257D%252C%2522clsw0ywvg00042v6h99qbxple%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clsw0ywvg00042v6h99qbxple%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D&workspaceId=58fd793c-a178-4c0e-8ac3-54df60751a6e)

**Comment by jbakarich** [comment]
Might not be the best way of going about it, but tweaking the `setFlip` function of `Navigator` to more closely match the logic in `Navigator.setRotation` seems to resolve the issue:

```js
    // $.Navigator.setFlip()
    /**
      * Flip navigator element
      * @param {Boolean} state - Flip state to set.
      */
    setFlip: function(state) {
      this.viewport.setFlip(state);

      var rule = this.viewer.viewport.getFlip() ? "scale(-1,1)" : "scale(1,1)"
      var displayRegionRule = "rotate(" + -this.viewport.getRotation() + "deg) " + rule;

      setElementTransform(this.canvas, rule);
      setElementTransform(this.element, rule);
      setElementTransform(this.displayRegion, displayRegionRule);

      return this;
    },
```

**Comment by iangilman** [comment]
Matching what we do for rotation seems like a reasonable approach. I'm not finding `Navigator.setRotation` in the code... what function are you referring to?

**Comment by jbakarich** [comment]
My mistake, its actually just called the `rotate` function found in `$.Navigator`: https://github.com/openseadragon/openseadragon/blob/e1e0ec6b241e11baec873a024ddf6a7e59834647/src/navigator.js#L223-L238

And what I mean by matching is that it seems that the existing flip code just overwrites the transform on the displayRegion div, we can just counter-rotate based on current rotation like is found on line 225:

https://github.com/openseadragon/openseadragon/blob/e1e0ec6b241e11baec873a024ddf6a7e59834647/src/navigator.js#L225

Thanks for all your work on OSD!

**Comment by iangilman** [comment]
I see! Seems like the best approach would be to have a common function that the rotate handler and setFlip both call, so either way it gets set properly (since either flip or rotate could be called first). What do you think? Would you be up for making a pull request?

**Comment by jbakarich** [comment]
Finally got some time to work on this! 

Have a quick question after looking into this a bit:

it seems that the main problem is that the main problem is that `setFlip()` is overriding existing rotation transforms for the display region when flipped:
https://github.com/openseadragon/openseadragon/blob/e1e0ec6b241e11baec873a024ddf6a7e59834647/src/navigator.js#L350C7-L350C26

What I don't quite understand though is that by flipping the element and canvas we are also flipping the display region, so I'm not sure that line is even needed. 

An example of what `setFlip()` appears to do (not including the `displayRegion` transform): 
[2024-11-20 13-52-25.webm](https://github.com/user-attachments/assets/e88243ea-a165-471b-a724-3cbc90061657)

I think that the `displayRegion` is correctly oriented just by the element/canvas flipping? 

Let me know if I'm incorrect, I can pursue just preserving the existing rotation transform while still flipping, or your suggestion of a shared method!

**Comment by iangilman** [comment]
Well spotted! Looks like the code evolved unevenly. Good to have that cleaned up!

Fixed by #2619.
